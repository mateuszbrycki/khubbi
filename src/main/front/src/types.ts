export interface Event {
    readonly note: string;
    readonly date: string;
    readonly id: string;
}

export interface User {
    readonly id: string;
    readonly email: string;
    readonly createdAt: string;
}

export class EventForms {
    public static readonly NOTE = new EventForms('Note');
    public static readonly PICTURE = new EventForms('Picture');
    public static readonly EVENT = new EventForms('Event');
    public static readonly EXPENSE = new EventForms('Expense');
    public static readonly REMINDER = new EventForms('Reminder');

    private constructor(public readonly name: string) {
    }
}