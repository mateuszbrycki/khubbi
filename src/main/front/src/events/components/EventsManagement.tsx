import {List} from "immutable"
import React from "react";
import Button from "react-bootstrap/Button";

import * as Icon from "react-bootstrap-icons";
import {Col, Row} from "react-bootstrap";
import Container from "react-bootstrap/Container";
import {EventDate, EventForms} from "../../types";
import AddNoteForm from "./AddNoteForm";
import AddPhotoForm from "./AddPhotoForm";

export interface EventsManagementProps {
    readonly showEventForm: EventForms | null
}

export interface EventsManagementActionProps {
    readonly openAddEventForm: (type: EventForms) => void,
    readonly closeAddEventForm: () => void,
    readonly addNote: (note: string, date: EventDate) => void
    readonly addPhoto: (description: string, photo: File, date: EventDate) => void
}

const EventsManagement: React.FC<EventsManagementProps & EventsManagementActionProps> = (props) => {

    const {showEventForm, openAddEventForm, closeAddEventForm, addNote, addPhoto} = props

    const buttons: List<JSX.Element> = List.of({
        eventForm: EventForms.NOTE,
        icon: <Icon.FileEarmarkPostFill className="me-2"/>
    }, {
        eventForm: EventForms.PICTURE,
        icon: <Icon.ImageFill className="me-2"/>
    }, {
        eventForm: EventForms.EVENT,
        icon: <Icon.CalendarEventFill className="me-2"/>
    }, {
        eventForm: EventForms.EXPENSE,
        icon: <Icon.WalletFill className="me-2"/>
    }, {
        eventForm: EventForms.REMINDER,
        icon: <Icon.StopwatchFill className="me-2"/>
    }).map(form => <Col><Button
        className={`btn-light d-inline m-3 ${showEventForm === form.eventForm ? " active" : ""}`}
        onClick={() => showEventForm === form.eventForm ? closeAddEventForm() : openAddEventForm(form.eventForm)}
        data-testid={`show-add-${form.eventForm.name}-button`}
    >{form.eventForm.name} {form.icon}</Button></Col>)

    const renderForm = (form: EventForms) => {
        switch (form) {
            case EventForms.NOTE:
                return <AddNoteForm
                    addNote={addNote}
                    currentDateProvider={() => EventDate.now()}/>
            case EventForms.PICTURE:
                return <AddPhotoForm
                    data-testid="addphoto-form"
                    addPhoto={addPhoto}
                    currentDateProvider={() => EventDate.now()}/>
            default:
                return <>No form yet</>
        }
    }

    return <>
        <Container className="justify-content-center">
            <Row>
                {buttons}
            </Row>

            <Row className="justify-content-center">
                <div className="col-6">
                    <Row className="justify-content-end mb-3">
                        {showEventForm !== null ?
                            <Icon.XCircle
                                data-testid="close-add-event-form-button"
                                className="col-1"
                                type="submit"
                                onClick={() => closeAddEventForm()}/>
                            : <></>}
                    </Row>
                    <Row>
                        {showEventForm !== null ? renderForm(showEventForm) : <></>}
                    </Row>
                </div>
            </Row>

        </Container>
    </>
}

export default EventsManagement