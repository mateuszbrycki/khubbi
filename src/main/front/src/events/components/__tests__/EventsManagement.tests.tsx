import ShallowRenderer from "react-test-renderer/shallow";
import EventsManagement from "../EventsManagement";
import {EventForms} from "../../../types";
import {act, fireEvent, render} from "@testing-library/react";

describe('renders management bar', () => {
    it('without any form open', () => {
        const renderer = ShallowRenderer.createRenderer()
        renderer.render(<EventsManagement
            showEventForm={null}
            openAddEventForm={() => null}
            closeAddEventForm={() => null}
            addNote={() => null}
            addPhoto={() => null}
        />)
        expect(renderer.getRenderOutput()).toMatchSnapshot()
    });

    it('with AddNote form open', () => {
        const renderer = ShallowRenderer.createRenderer()
        renderer.render(<EventsManagement
            showEventForm={EventForms.NOTE}
            openAddEventForm={() => null}
            closeAddEventForm={() => null}
            addNote={() => null}
            addPhoto={() => null}
        />)
        expect(renderer.getRenderOutput()).toMatchSnapshot()
    });

    it('with AddPhoto form open', () => {
        const renderer = ShallowRenderer.createRenderer()
        renderer.render(<EventsManagement
            showEventForm={EventForms.PICTURE}
            openAddEventForm={() => null}
            closeAddEventForm={() => null}
            addNote={() => null}
            addPhoto={() => null}
        />)
        expect(renderer.getRenderOutput()).toMatchSnapshot()
    });
})

test("should open the form after clicking at the button and mark it active", async () => {
    const showEventFormMock = jest.fn();

    const {getByTestId} = render(<EventsManagement
        showEventForm={null}
        openAddEventForm={showEventFormMock}
        closeAddEventForm={() => null}
        addNote={() => null}
        addPhoto={() => null}
    />);

    await act(async () => {
        getByTestId("show-add-Note-button").click();
    })

    expect(showEventFormMock).toHaveBeenCalledWith(EventForms.NOTE);
})

test("button for open form for should have `active` class", async () => {

    const {getByTestId} = render(<EventsManagement
        showEventForm={EventForms.NOTE}
        openAddEventForm={() => null}
        closeAddEventForm={() => null}
        addNote={() => null}
        addPhoto={() => null}
    />);

    expect(getByTestId("show-add-Note-button").getAttribute("class")).toContain("active")

})

test("should close the form after clicking at the button", async () => {
    const closeAddEventFormMock = jest.fn();

    const {getByTestId} = render(<EventsManagement
        showEventForm={EventForms.NOTE}
        openAddEventForm={() => null}
        closeAddEventForm={closeAddEventFormMock}
        addNote={() => null}
        addPhoto={() => null}
    />);

    await act(async () => {
        getByTestId("show-add-Note-button").click();
    })

    expect(closeAddEventFormMock).toHaveBeenCalled();
})
test("should close the form after clicking at the close button", async () => {
    const closeAddEventFormMock = jest.fn();

    const {getByTestId} = render(<EventsManagement
        showEventForm={EventForms.NOTE}
        openAddEventForm={() => null}
        closeAddEventForm={closeAddEventFormMock}
        addNote={() => null}
        addPhoto={() => null}
    />);

    await act(async () => {
        fireEvent.click(getByTestId("close-add-event-form-button"))
    })

    expect(closeAddEventFormMock).toHaveBeenCalled();
})