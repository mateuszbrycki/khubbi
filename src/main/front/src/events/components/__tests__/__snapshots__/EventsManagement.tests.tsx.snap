// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders management bar with AddNote form open 1`] = `
<React.Fragment>
  <Container
    className="justify-content-center"
  >
    <Row>
      Immutable.List [
        <Col>
          <Button
            className="btn-light d-inline m-3  active"
            data-testid="show-add-Note-button"
            onClick={[Function]}
          >
            Note
             
            <ForwardRef
              className="me-2"
              color="currentColor"
              size="1em"
              title={null}
            />
          </Button>
        </Col>,
        <Col>
          <Button
            className="btn-light d-inline m-3 "
            data-testid="show-add-Picture-button"
            onClick={[Function]}
          >
            Picture
             
            <ForwardRef
              className="me-2"
              color="currentColor"
              size="1em"
              title={null}
            />
          </Button>
        </Col>,
        <Col>
          <Button
            className="btn-light d-inline m-3 "
            data-testid="show-add-Event-button"
            onClick={[Function]}
          >
            Event
             
            <ForwardRef
              className="me-2"
              color="currentColor"
              size="1em"
              title={null}
            />
          </Button>
        </Col>,
        <Col>
          <Button
            className="btn-light d-inline m-3 "
            data-testid="show-add-Expense-button"
            onClick={[Function]}
          >
            Expense
             
            <ForwardRef
              className="me-2"
              color="currentColor"
              size="1em"
              title={null}
            />
          </Button>
        </Col>,
        <Col>
          <Button
            className="btn-light d-inline m-3 "
            data-testid="show-add-Reminder-button"
            onClick={[Function]}
          >
            Reminder
             
            <ForwardRef
              className="me-2"
              color="currentColor"
              size="1em"
              title={null}
            />
          </Button>
        </Col>,
      ]
    </Row>
    <Row
      className="justify-content-center"
    >
      <div
        className="col-6"
      >
        <Row
          className="justify-content-end mb-3"
        >
          <ForwardRef
            className="col-1"
            color="currentColor"
            data-testid="close-add-event-form-button"
            onClick={[Function]}
            size="1em"
            title={null}
            type="submit"
          />
        </Row>
        <Row>
          <AddNoteForm
            addNote={[Function]}
            currentDateProvider={[Function]}
          />
        </Row>
      </div>
    </Row>
  </Container>
</React.Fragment>
`;

exports[`renders management bar without any form open 1`] = `
<React.Fragment>
  <Container
    className="justify-content-center"
  >
    <Row>
      Immutable.List [
        <Col>
          <Button
            className="btn-light d-inline m-3 "
            data-testid="show-add-Note-button"
            onClick={[Function]}
          >
            Note
             
            <ForwardRef
              className="me-2"
              color="currentColor"
              size="1em"
              title={null}
            />
          </Button>
        </Col>,
        <Col>
          <Button
            className="btn-light d-inline m-3 "
            data-testid="show-add-Picture-button"
            onClick={[Function]}
          >
            Picture
             
            <ForwardRef
              className="me-2"
              color="currentColor"
              size="1em"
              title={null}
            />
          </Button>
        </Col>,
        <Col>
          <Button
            className="btn-light d-inline m-3 "
            data-testid="show-add-Event-button"
            onClick={[Function]}
          >
            Event
             
            <ForwardRef
              className="me-2"
              color="currentColor"
              size="1em"
              title={null}
            />
          </Button>
        </Col>,
        <Col>
          <Button
            className="btn-light d-inline m-3 "
            data-testid="show-add-Expense-button"
            onClick={[Function]}
          >
            Expense
             
            <ForwardRef
              className="me-2"
              color="currentColor"
              size="1em"
              title={null}
            />
          </Button>
        </Col>,
        <Col>
          <Button
            className="btn-light d-inline m-3 "
            data-testid="show-add-Reminder-button"
            onClick={[Function]}
          >
            Reminder
             
            <ForwardRef
              className="me-2"
              color="currentColor"
              size="1em"
              title={null}
            />
          </Button>
        </Col>,
      ]
    </Row>
    <Row
      className="justify-content-center"
    >
      <div
        className="col-6"
      >
        <Row
          className="justify-content-end mb-3"
        >
          <React.Fragment />
        </Row>
        <Row>
          <React.Fragment />
        </Row>
      </div>
    </Row>
  </Container>
</React.Fragment>
`;
